File: examples/typescript/vrf_example.ts
/**
 * VRF (Verifiable Random Function) Service Example
 *
 * This example demonstrates how to generate verifiable random numbers
 * for fair selection, lotteries, and gaming applications.
 *
 * Run: npx ts-node vrf_example.ts
 */

import { createClient, SupabaseClient } from '@supabase/supabase-js';

// =============================================================================
// Configuration
// =============================================================================

const SUPABASE_URL = process.env.VITE_SUPABASE_URL || 'http://localhost:54321';
const SUPABASE_KEY = process.env.VITE_SUPABASE_ANON_KEY || '';

// =============================================================================
// Types
// =============================================================================

interface ServiceRequest {
  id?: string;
  service_type: string;
  operation: string;
  payload: Record<string, unknown>;
  status?: string;
  result?: Record<string, unknown>;
  error_message?: string;
}

interface VRFResult {
  randomness: string;
  proof: string;
  seed: string;
}

// =============================================================================
// VRF Service Client
// =============================================================================

class VRFService {
  private supabase: SupabaseClient;

  constructor(url: string, key: string) {
    this.supabase = createClient(url, key);
  }

  /**
   * Generate a verifiable random number
   */
  async generateRandom(seed: string, numValues: number = 1): Promise<VRFResult> {
    // Submit request
    const { data: request, error } = await this.supabase
      .from('service_requests')
      .insert({
        service_type: 'vrf',
        operation: 'random',
        payload: { seed, num_values: numValues },
      })
      .select()
      .single();

    if (error) throw new Error(`Submit failed: ${error.message}`);

    // Wait for completion
    const result = await this.waitForCompletion(request.id);

    if (result.status === 'failed') {
      throw new Error(result.error_message || 'VRF generation failed');
    }

    return result.result as VRFResult;
  }

  /**
   * Verify a VRF proof
   */
  async verifyProof(randomness: string, proof: string, seed: string): Promise<boolean> {
    const { data: request, error } = await this.supabase
      .from('service_requests')
      .insert({
        service_type: 'vrf',
        operation: 'verify',
        payload: { randomness, proof, seed },
      })
      .select()
      .single();

    if (error) throw new Error(`Submit failed: ${error.message}`);

    const result = await this.waitForCompletion(request.id);
    return result.result?.valid === true;
  }

  private async waitForCompletion(id: string, timeoutMs: number = 30000): Promise<ServiceRequest> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeoutMs) {
      const { data } = await this.supabase
        .from('service_requests')
        .select('*')
        .eq('id', id)
        .single();

      if (data?.status === 'completed' || data?.status === 'failed') {
        return data;
      }

      await new Promise(resolve => setTimeout(resolve, 1000));
    }

    throw new Error('Timeout');
  }
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Convert hex randomness to a number in range [0, max)
 */
function randomToRange(randomnessHex: string, max: number): number {
  const randomBigInt = BigInt('0x' + randomnessHex);
  return Number(randomBigInt % BigInt(max));
}

/**
 * Shuffle array using VRF randomness (Fisher-Yates)
 */
function shuffleWithRandomness(array: string[], randomnessHex: string): string[] {
  const result = [...array];
  const randomBigInt = BigInt('0x' + randomnessHex);

  for (let i = result.length - 1; i > 0; i--) {
    const j = Number((randomBigInt >> BigInt(i * 8)) % BigInt(i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }

  return result;
}

// =============================================================================
// Examples
// =============================================================================

async function main() {
  console.log('=== VRF Service Example ===\n');

  if (!SUPABASE_KEY) {
    console.error('Error: VITE_SUPABASE_ANON_KEY environment variable required');
    process.exit(1);
  }

  const vrf = new VRFService(SUPABASE_URL, SUPABASE_KEY);

  // Example 1: Simple random number
  console.log('Example 1: Generate verifiable random number');
  console.log('-'.repeat(50));

  try {
    const seed = `lottery-${Date.now()}`;
    console.log(`Seed: ${seed}`);

    const result = await vrf.generateRandom(seed);

    console.log('✓ Random number generated!');
    console.log(`  Randomness: ${result.randomness.substring(0, 32)}...`);
    console.log(`  Proof: ${result.proof.substring(0, 32)}...`);
    console.log();
  } catch (error) {
    console.error('Error:', error);
  }

  // Example 2: Fair lottery winner selection
  console.log('Example 2: Fair lottery winner selection');
  console.log('-'.repeat(50));

  try {
    const participants = ['Alice', 'Bob', 'Charlie', 'Diana', 'Eve', 'Frank'];
    console.log(`Participants: ${participants.join(', ')}`);

    const seed = `lottery-winner-${Date.now()}`;
    const result = await vrf.generateRandom(seed);

    const winnerIndex = randomToRange(result.randomness, participants.length);
    const winner = participants[winnerIndex];

    console.log('✓ Winner selected!');
    console.log(`  Winner: ${winner} (index ${winnerIndex})`);
    console.log(`  Seed: ${seed}`);
    console.log(`  Proof available for verification`);
    console.log();
  } catch (error) {
    console.error('Error:', error);
  }

  // Example 3: NFT trait randomization
  console.log('Example 3: NFT trait randomization');
  console.log('-'.repeat(50));

  try {
    const traits = {
      background: ['Blue', 'Red', 'Green', 'Purple', 'Gold'],
      body: ['Robot', 'Human', 'Alien', 'Zombie'],
      accessory: ['Hat', 'Glasses', 'Earring', 'None'],
    };

    const tokenId = 12345;
    const seed = `nft-traits-${tokenId}`;
    const result = await vrf.generateRandom(seed);

    const randomBigInt = BigInt('0x' + result.randomness);

    const selectedTraits = {
      background: traits.background[Number(randomBigInt % BigInt(traits.background.length))],
      body: traits.body[Number((randomBigInt >> 8n) % BigInt(traits.body.length))],
      accessory: traits.accessory[Number((randomBigInt >> 16n) % BigInt(traits.accessory.length))],
    };

    console.log(`✓ NFT #${tokenId} traits generated!`);
    console.log(`  Background: ${selectedTraits.background}`);
    console.log(`  Body: ${selectedTraits.body}`);
    console.log(`  Accessory: ${selectedTraits.accessory}`);
    console.log();
  } catch (error) {
    console.error('Error:', error);
  }

  // Example 4: Dice roll for gaming
  console.log('Example 4: Provably fair dice roll');
  console.log('-'.repeat(50));

  try {
    const gameId = 'game-' + Date.now();
    const seed = `dice-roll-${gameId}`;
    const result = await vrf.generateRandom(seed);

    const diceValue = randomToRange(result.randomness, 6) + 1; // 1-6

    console.log(`✓ Dice rolled!`);
    console.log(`  Value: ${diceValue}`);
    console.log(`  Game ID: ${gameId}`);
    console.log(`  Verifiable: Anyone can verify with seed + proof`);
    console.log();
  } catch (error) {
    console.error('Error:', error);
  }

  console.log('=== Example Complete ===');
}

// Run
main().catch(console.error);

export { VRFService, randomToRange, shuffleWithRandomness };

File: examples/go/vrf_example.go
// Package main demonstrates how to use the VRF (Verifiable Random Function) service.
//
// This example shows:
// - Generating verifiable random numbers
// - Verifying VRF proofs
// - Using randomness for fair selection
//
// Run: go run vrf_example.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math/big"
	"net/http"
	"os"
	"time"
)

// =============================================================================
// Configuration
// =============================================================================

func getEnv(key, defaultValue string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return defaultValue
}

// =============================================================================
// Types
// =============================================================================

type ServiceRequest struct {
	ID          string                 `json:"id,omitempty"`
	ServiceType string                 `json:"service_type"`
	Operation   string                 `json:"operation"`
	Payload     map[string]interface{} `json:"payload"`
	Status      string                 `json:"status,omitempty"`
	Result      map[string]interface{} `json:"result,omitempty"`
	Error       string                 `json:"error_message,omitempty"`
}

// =============================================================================
// Client
// =============================================================================

type Client struct {
	url    string
	key    string
	client *http.Client
}

func NewClient(url, key string) *Client {
	return &Client{
		url:    url,
		key:    key,
		client: &http.Client{Timeout: 30 * time.Second},
	}
}

func (c *Client) Submit(ctx context.Context, req ServiceRequest) (*ServiceRequest, error) {
	body, _ := json.Marshal(req)
	httpReq, _ := http.NewRequestWithContext(ctx, "POST",
		c.url+"/rest/v1/service_requests", bytes.NewReader(body))

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("apikey", c.key)
	httpReq.Header.Set("Authorization", "Bearer "+c.key)
	httpReq.Header.Set("Prefer", "return=representation")

	resp, err := c.client.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusCreated {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed: %s", body)
	}

	var results []ServiceRequest
	json.NewDecoder(resp.Body).Decode(&results)
	return &results[0], nil
}

func (c *Client) Wait(ctx context.Context, id string) (*ServiceRequest, error) {
	for i := 0; i < 30; i++ {
		httpReq, _ := http.NewRequestWithContext(ctx, "GET",
			c.url+"/rest/v1/service_requests?id=eq."+id, nil)
		httpReq.Header.Set("apikey", c.key)
		httpReq.Header.Set("Authorization", "Bearer "+c.key)

		resp, err := c.client.Do(httpReq)
		if err != nil {
			return nil, err
		}

		var results []ServiceRequest
		json.NewDecoder(resp.Body).Decode(&results)
		resp.Body.Close()

		if len(results) > 0 && (results[0].Status == "completed" || results[0].Status == "failed") {
			return &results[0], nil
		}

		time.Sleep(1 * time.Second)
	}
	return nil, fmt.Errorf("timeout")
}

// =============================================================================
// Main
// =============================================================================

func main() {
	log.Println("=== VRF Service Example ===")
	log.Println()

	url := getEnv("SUPABASE_URL", "http://localhost:54321")
	key := getEnv("SUPABASE_ANON_KEY", "")
	if key == "" {
		log.Fatal("SUPABASE_ANON_KEY required")
	}

	client := NewClient(url, key)
	ctx := context.Background()

	// Example 1: Generate random number
	log.Println("Example 1: Generating verifiable random number...")

	req := ServiceRequest{
		ServiceType: "vrf",
		Operation:   "random",
		Payload: map[string]interface{}{
			"seed": "lottery-round-12345",
		},
	}

	submitted, err := client.Submit(ctx, req)
	if err != nil {
		log.Fatalf("Submit error: %v", err)
	}

	log.Printf("Request ID: %s", submitted.ID)

	result, err := client.Wait(ctx, submitted.ID)
	if err != nil {
		log.Fatalf("Wait error: %v", err)
	}

	if result.Status == "completed" {
		log.Println("✓ Random number generated!")
		log.Printf("  Randomness: %v", result.Result["randomness"])
		log.Printf("  Proof: %v", result.Result["proof"])
	} else {
		log.Printf("✗ Failed: %s", result.Error)
	}

	log.Println()

	// Example 2: Fair lottery selection
	log.Println("Example 2: Fair lottery winner selection...")

	participants := []string{"Alice", "Bob", "Charlie", "Diana", "Eve"}
	log.Printf("Participants: %v", participants)

	req2 := ServiceRequest{
		ServiceType: "vrf",
		Operation:   "random",
		Payload: map[string]interface{}{
			"seed": fmt.Sprintf("lottery-%d", time.Now().UnixNano()),
		},
	}

	submitted2, _ := client.Submit(ctx, req2)
	result2, _ := client.Wait(ctx, submitted2.ID)

	if result2.Status == "completed" {
		// Use randomness to select winner
		if randomHex, ok := result2.Result["randomness"].(string); ok {
			randomBig := new(big.Int)
			randomBig.SetString(randomHex, 16)

			winnerIndex := new(big.Int).Mod(randomBig, big.NewInt(int64(len(participants))))
			winner := participants[winnerIndex.Int64()]

			log.Printf("✓ Winner: %s (index %d)", winner, winnerIndex.Int64())
			log.Printf("  Proof available for verification")
		}
	}

	log.Println()

	// Example 3: Multiple random values
	log.Println("Example 3: Generating multiple random values...")

	req3 := ServiceRequest{
		ServiceType: "vrf",
		Operation:   "random",
		Payload: map[string]interface{}{
			"seed":       "multi-random-seed",
			"num_values": 5,
		},
	}

	submitted3, _ := client.Submit(ctx, req3)
	result3, _ := client.Wait(ctx, submitted3.ID)

	if result3.Status == "completed" {
		log.Println("✓ Multiple random values generated!")
		if values, ok := result3.Result["values"].([]interface{}); ok {
			for i, v := range values {
				log.Printf("  Value %d: %v", i+1, v)
			}
		}
	}

	log.Println()
	log.Println("=== Example Complete ===")
}
